Code/GraphMol/CIPLabeler/CIPLabeler.cpp

```diff
@@ -0,0 +1,203 @@
+//
+//
+//  Copyright (C) 2020 Schrödinger, LLC
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include <algorithm>
+
+#include <GraphMol/RDKitBase.h>
+
+#include "CIPLabeler.h"
+#include "CIPMol.h"
+#include "configs/Sp2Bond.h"
+#include "configs/Tetrahedral.h"
+
+#include "rules/Rules.hpp"
+#include "rules/Rule1a.h"
+#include "rules/Rule1b.h"
+#include "rules/Rule2.h"
+#include "rules/Rule3.h"
+#include "rules/Rule4a.h"
+#include "rules/Rule4b.h"
+#include "rules/Rule4c.h"
+#include "rules/Rule5New.h"
+#include "rules/Rule6.h"
+
+namespace RDKit {
+namespace CIPLabeler {
+
+namespace {
+std::vector<Configuration *> findConfigs(CIPMol &mol) {
+  std::vector<Configuration *> configs;
+
+  for (auto &atom : mol.atoms()) {
+    auto chiraltag = static_cast<int>(atom->getChiralTag());
+    if (chiraltag > 0) {
+      auto carriers = std::vector<Atom *>();
+      carriers.reserve(4);
+      for (auto &bond : mol.getBonds(atom)) {
+        auto nbr = bond->getOtherAtom(atom);
+        carriers.push_back(nbr);
+      }
+      if (carriers.size() < 4) {
+        // Implicit H -- use the central atom instead of a dummy H
+        carriers.push_back(atom);
+      }
+
+      // RDKit uses SMILES parity, and centres requires SDF parity, which
+      // is opposite.
+      auto parity = chiraltag ^ 0x3;
+
+      auto cfg = new Tetrahedral(atom, std::move(carriers), parity);
+      configs.push_back(cfg);
+    }
+  }
+
+  for (auto i = 0u; i < mol.getNumBonds(); ++i) {
+    auto bond = mol.getBond(i);
+
+    if (bond->getBondType() == Bond::DOUBLE) {
+      std::vector<Atom *> atoms{{bond->getBeginAtom(), bond->getEndAtom()}};
+      std::vector<Atom *> anchors;
+      auto previous_bond_dir = Bond::BondDir::NONE;
+      auto bond_cfg = 1;
+      for (const auto &atom : atoms) {
+        for (auto &nbr_bnd : mol.getBonds(atom)) {
+          auto nbr_bond_dir = nbr_bnd->getBondDir();
+          if (nbr_bond_dir == Bond::BondDir::ENDDOWNRIGHT ||
```

Comment:
I don't know if this is the canonical way to check for bond stereo. I just wanted to go to the lowest possible level of information, as having CIS/TRANS or E/Z on the molecule implies that some processing has already been done.

Reply:
I think it's worth requiring that CIS/TRANS be assigned, anything else is too complicated.

Reply:
That's also interesing, since for some cases we'd need to translate from E/Z, which is kind of ironic, since that would mean we already have calculated stereo on that bond at some point.

Reply:
yeah, but we're using E/Z as cis/trans.

Reply:
Right, but E/Z have the disadvantage that StereoAtoms are not set, which is required for cis/trans.